ConfigSpace/forbidden.pyx

```diff
@@ -564,13 +568,74 @@ cdef class ForbiddenRelation(AbstractForbiddenComponent):
             else:
                 return False
 
-        # Relation is always evaluated against actual value and not vector representation
-        return self._is_forbidden(self.left._transform(left), self.right._transform(right))
+        # Relation is always evaluated against actual value and
+        # not vector representation
+        return self._is_forbidden(self.left._transform(left),
+                                  self.right._transform(right))
 
     cdef int _is_forbidden_vector(self, DTYPE_t left, DTYPE_t right) except -1:
         pass
 
 
+cdef class ForbiddenCallableRelation(ForbiddenRelation):
+    """A ForbiddenCallable relation between two hyperparameters.
+
+    The ForbiddenCallable uses two hyperparameters as input to a
+    specified callable, which returns True if the relationship
+    between the two hyperparameters is forbidden.
+
+    >>> from ConfigSpace import ConfigurationSpace, ForbiddenLessThanRelation
+    >>>
+    >>> cs = ConfigurationSpace({"a": [1, 2, 3], "b": [2, 5, 6]})
+    >>>
+    >>> forbidden_clause = ForbiddenCallableRelation(cs['a'], cs['b'])
+    >>> cs.add_forbidden_clause(forbidden_clause)
+    Forbidden: f(a,b) == True
+
+    Parameters
+    ----------
+     left : :ref:`Hyperparameters`
+         first argument of callable
+
+     right : :ref:`Hyperparameters`
+         second argument of callable
+
+     f : A callable that relates the two hyperparameters
+    """
+    cdef public f
+
+    def __init__(self, left: Hyperparameter, right: Hyperparameter,
+                 f: ForbiddenCallable):
+        if not isinstance(f, Callable):  # Can't use ForbiddenCallable here apparently
+            raise TypeError("Argument 'f' is not of type %s." % Callable)
+
+        super().__init__(left, right)
+        self.f = f
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+        return super().__eq__(other) and self.f == other.f
+
+    def __copy__(self):
+        return self.__class__(
+            a=copy.copy(self.left),
+            b=copy.copy(self.right),
+            f=copy.copy(self.f)
+        )
+
+    def __repr__(self):
+        from inspect import getsource
+        f_source = getsource(self.f)
+        return f"Forbidden: {f_source} | Arguments: {self.left.name}, {self.right.name}"
+
+    cdef int _is_forbidden(self, left, right) except -1:
```

Comment:
What does this `except -1` do? I wish I knew more Cython but unfortunatly not. I looked at the class above and it didn't seem to have this.

Reply:
I modeled this off of the `ForbiddenLessThanRelation` directly below which uses this. I'm not very familiar with Cython either, but it seems any cdef function that might return a Python exception needs to be declared with an except value. https://docs.cython.org/en/latest/src/userguide/language_basics.html#error-return-values